include/mesh/dim2/adjacency_map/adjacency_map.hpp

```
@@ -0,0 +1,213 @@
+#pragma once
+
+#include "Kokkos_Macros.hpp"
+#include "enumerations/dimension.hpp"
+#include "enumerations/specfem_enums.hpp"
+#include <Kokkos_Core.hpp>
+#include <set>
+
+namespace specfem {
+namespace mesh {
+template <specfem::dimension::type Dimension> struct mesh;
+namespace adjacency_map {
+
+/**
+ * @brief Stores the adjacencies between elements.
+ *
+ */
+template <specfem::dimension::type Dimension> struct adjacency_map;
+template <> struct adjacency_map<specfem::dimension::type::dim2> {
+
+  /**
+   * @brief an enumeration of the edges used by adjacency_map. Recovers the
+   * index from an edge::type -- this is set explicitly since it must not
+   * change, and some logic is simplified with this counter-clockwise ordering.
+   */
+  static constexpr int edge_to_index(const specfem::enums::edge::type edge) {
+    switch (edge) {
+    case specfem::enums::edge::RIGHT:
+      return 0;
+    case specfem::enums::edge::TOP:
+      return 1;
+    case specfem::enums::edge::LEFT:
+      return 2;
+    case specfem::enums::edge::BOTTOM:
+      return 3;
+    default:
```

You should throw an error if the behavior is unexpected.